---
title: "testing"
author: "Wes Horton"
date: "July 14, 2016"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
spikes <- read.table("~/Desktop/OHSU/tcr_spike/text_barcodesvj.txt", header = T, sep = ' ', stringsAsFactors = F)
library(ggplot2)
library(reshape2)
```

### Overview

When testing primer independence using samples 1-20 from DNA160609LC, Burcu noticed a four-fold range in spike counts between samples. For example, the spike corresponding to V1/J1-1 in sample 1 could have a count of 10, whereas the same spike has a count of 40 in sample 2. We usually give the sequencing core unequal amounts of DNA in our samples, but these 20 samples have the same initial concentration of spikes. We would expect them to then have the same final concentration as well, assuming that they amplified similarly.

We can compare the 9-bp spike count totals with the PEAR fastq read counts and determine if the distributions are similar.


```{r, echo = F}
### 9-bp counts
spike.count.dir <- "/Volumes/DNA160609LC/spike_counts/9bp/counts/"
spike.count.files <- list.files(spike.count.dir)
spike.count.files <- spike.count.files[order(as.numeric(gsub(".*_S|\\.assemb.*", '', spike.count.files)))]
spike.count.files <- spike.count.files[1:20]

### PEAR files
### Read in and sort pear files
#pear.dir <- "/Volumes/DNA160609LC/peared_fastqs/counting/"
#pear.files <- list.files(pear.dir)
#pear.files <- pear.files[order(as.numeric(gsub(".*_S|\\.assem.*", '', pear.files)))]
```

#### Compare distributions
```{r, echo = F}
# Spike counts of 9-bp counts
spike.counts <- NULL
for (i in 1:length(spike.count.files)){
  curr.spike.count <- read.table(paste(spike.count.dir, spike.count.files[i], sep = ''),
                                 header = T, stringsAsFactors = F, sep = ',')
  curr.count <- curr.spike.count$spike.count[1]
  spike.counts <- c(spike.counts, curr.count)
} # for

# Read counts of raw files
# pear.counts <- NULL
# for (i in 1:length(pear.files)){
#   curr.pear.count <- as.numeric(system(paste("grep -c '^@' ", pear.dir, pear.files[i], sep = ''), intern = T))
#   pear.counts <- c(pear.counts, curr.pear.count)
# } # for
# Above takes too long, just did in bash and imported:
pear.counts <- c(686374, 1209673,  865117, 1095477, 1445022, 1100437, 1433990, 1864230, 
                 1328501, 1078247, 1916805, 1011097,  768413,  508421,  925440,  716742, 572435, 914924,  
                 723413, 1169669)


count.compare <- data.frame("Sample" = seq(1:20), "Spike.counts" = spike.counts, "Pear.counts" = pear.counts)
count.compare

count.compare.melt <- melt(count.compare, id.vars = "Sample", variable.name = "Count.type", value.name = "Count")

gg.count.compare <- ggplot(count.compare.melt, aes(x = Count.type, y = Count)) +
  ggtitle("Comparison of raw and spike counts")
gg.count.compare + geom_point()
```

We can see that the distributions are almost identical. These results suggest that the variation came from any of the following (or a combination of them):

1. Sample preparation - variable amounts of spikes were pipetted into the samples
2. PCR amplification - samples have same starting material, but individual reactions amplified differently in the thermocycler
3. Sequencing - Samples seeded unevenly on the sequencing machine

### Next Steps

Do we have quantitative data of sample concentration prior to PCR (nanodrop or cubit?), if so, we should check that distribution. We can also ask Bob if he saw a difference in seeds.

### Compare with PCR Concentrations

Is there a correlation between pcr concentration and spike counts

```{r, echo = F}
# Read in PCR data
pcr.conc.data <- read.csv("~/Desktop/OHSU/tcr_spike/data/DNA160609LC/spike_only_pcr.csv",
                          stringsAsFactors = F)

# Make data frames
# PCR1 Compare
pcr1.counts <- cbind(pcr.conc.data[,c(1,2)], spike.counts)
# PCR2 Compare
pcr2.counts <- cbind(pcr.conc.data[,c(1,3)], spike.counts)

# Make Models
pcr1.model <- lm(spike.counts ~ PCR1, data = pcr1.counts)
summary(pcr1.model)

pcr2.model <- lm(spike.counts ~ PCR2, data = pcr2.counts)
summary(pcr2.model)


#Plots
# PCR1
pcr1.plot <- ggplot(pcr1.counts, aes(x = PCR1, y = spike.counts)) + geom_point() +
  ggtitle("Sample Concentration after PCR1 \nvs. Spike Counts") +
  xlab("DNA Concentration (ng/ul)") +
  ylab("Spike Counts") +
  ylim(min(pcr1.counts$spike.counts) *.9, max(pcr1.counts$spike.counts)* 1.1)
pcr1.plot + stat_smooth(method = lm, se = F)
# PCR2
pcr2.plot <- ggplot(pcr2.counts, aes(x = PCR2, y = spike.counts)) + geom_point() +
    ggtitle("Sample Concentration after PCR2 \nvs. Spike Counts") +
  xlab("DNA Concentration (ng/ul)") +
  ylab("Spike Counts") +
  ylim(min(pcr1.counts$spike.counts) *.9, max(pcr1.counts$spike.counts)* 1.1)
pcr2.plot + stat_smooth(method = lm, se = F)




# Log stuff
# PCR 1
# data
pcr1.log.counts <- cbind(pcr.conc.data[,c(1,2)], "spike.counts" = log2(spike.counts))
pcr1.log.counts$PCR1 <- log2(pcr1.log.counts$PCR1)
# Model
pcr1.log.model <- lm(spike.counts ~ PCR1, data = pcr1.log.counts)
summary(pcr1.log.model)
# Plot
pcr1.log.plot <- ggplot(pcr1.log.counts, aes(x = PCR1, y = spike.counts)) + geom_point() +
  ggtitle("Sample Concentration after PCR1 \nvs. Spike Counts") +
  xlab("log2(DNA Concentration) (ng/ul)") +
  ylab("log2(Spike Counts)") +
  ylim(min(pcr1.log.counts$spike.counts) *.9, max(pcr1.log.counts$spike.counts)* 1.1) +
  annotate("text", x = 4.8, y = 22, label = paste("Adj. R2: ", round(summary(pcr1.log.model)$adj.r.squared, digits = 3)))
pcr1.log.plot + stat_smooth(method = lm, se = F)

#PCR1
# data
pcr2.log.counts <- cbind(pcr.conc.data[,c(1,3)], "spike.counts" = log2(spike.counts))
pcr2.log.counts$PCR2 <- log2(pcr2.log.counts$PCR2)
# Model
pcr2.log.model <- lm(spike.counts ~ PCR2, data = pcr2.log.counts)
summary(pcr2.log.model)
# Plot
pcr2.log.plot <- ggplot(pcr2.log.counts, aes(x = PCR2, y = spike.counts)) + geom_point() +
  ggtitle("Sample Concentration after PCR2 \nvs. Spike Counts") +
  xlab("log2(DNA Concentration) (ng/ul)") +
  ylab("log2(Spike Counts)") +
  ylim(min(pcr2.log.counts$spike.counts) *.9, max(pcr2.log.counts$spike.counts)* 1.1) +
  annotate("text", x = 3.5, y = 22, label = paste("Adj. R2: ", round(summary(pcr2.log.model)$adj.r.squared, digits = 3)))
pcr2.log.plot + stat_smooth(method = lm, se = F)




```

Looks like the pcr concentrations at the end of the second round correlate well with the spike counts observed in the samples.

```{r, echo = F}
# Variation between PCR1 and PCR2
summary(pcr.conc.data$PCR1)
mean(pcr.conc.data$PCR1)
sd(pcr.conc.data$PCR1)
range(pcr.conc.data$PCR1)[2] - range(pcr.conc.data$PCR1)[1]

summary(pcr.conc.data$PCR2)
mean(pcr.conc.data$PCR2)
sd(pcr.conc.data$PCR2)
range(pcr.conc.data$PCR2)[2] - range(pcr.conc.data$PCR2)[1]

# Normalize by dividing by average
pcr.conc.data$norm.pcr1 <- pcr.conc.data$PCR1 / mean(pcr.conc.data$PCR1)
pcr.conc.data$norm.pcr2 <- pcr.conc.data$PCR2 / mean(pcr.conc.data$PCR2)

summary(pcr.conc.data$norm.pcr1)
summary(pcr.conc.data$norm.pcr2)


```

Much higher variation in the second 